router/dataplane_test.go

```diff
@@ -49,6 +51,175 @@ import (
+			t.Fail()
+		}
+	}
+}
+
+func TestForwarder(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	done := make(chan struct{})
+	prepareDP := func(ctrl *gomock.Controller) *router.DataPlane {
+		ret := &router.DataPlane{Metrics: metrics}
+
+		key := []byte("testkey_xxxxxxxx")
+		local := xtest.MustParseIA("1-ff00:0:110")
+		mInternal := mock_router.NewMockBatchConn(ctrl)
+		matchFlags := gomock.Eq(syscall.MSG_DONTWAIT)
+		totalCount := 10
+		for i := 0; i < 10; i++ {
+			mInternal.EXPECT().WriteBatch(gomock.Any(), matchFlags).DoAndReturn(
+				func(ms underlayconn.Messages, flags int) (int, error) {
+					if totalCount == 0 {
+						return 0, nil
+					}
+					totalCount -= len(ms)
+					if totalCount == 0 {
+						done <- struct{}{}
+					}
+					return len(ms), nil
+				}).AnyTimes()
+		}
+		_ = ret.AddInternalInterface(mInternal, net.IP{})
+		_ = ret.SetIA(local)
+		_ = ret.SetKey(key)
+		return ret
+	}
+	dp := prepareDP(ctrl)
+	dp.InitializePacketPool(100)
+	dp.ConfigureBatchSize(64)
+	ni := router.NetworkInterface{InterfaceId: 0, Addr: nil, Conn: dp.GetInternalInterface()}
+	ch := dp.ConfigureForwarder(ni)
+	dp.SetRunning(true)
+	dp.InitMetrics()
+	defer dp.SetRunning(false)
+	go func() {
+		dp.InitForwarder(ni)
+	}()
+	assert.Equal(t, 100, dp.CurrentPoolSize())
+	for i := 0; i < 10; i++ {
+		pkt := dp.GetBufferFromPool()
+		assert.NotEqual(t, 100, dp.CurrentPoolSize())
+		pkt.UpdateFields(nil, nil, ni.InterfaceId)
+		dp.SendPacketToChannel(pkt, ch)
+	}
+	select {
+	case <-done:
+		assert.Equal(t, 100, dp.CurrentPoolSize())
+	case <-time.After(100 * time.Millisecond):
+		t.Fail()
+	}
+}
+
+func TestComputeProcId(t *testing.T) {
+	packetBuffer := make([]byte, 9000)
+	dp := &router.DataPlane{}
+	randomValue := []byte{1, 2, 3, 4}
+	numProcs := 10000
+	dp.SetRandomValue(randomValue)
+	dp.ConfigureProcChannels(numProcs, 1)
+
+	key := []byte("testkey_xxxxxxxx")
+	local := xtest.MustParseIA("1-ff00:0:110")
+	spkt, dpath := prepBaseMsg(time.Now())
+	spkt.DstIA = local
+	spkt.FlowID = (1 << 20) - 1
+	dpath.HopFields = []path.HopField{
+		{ConsIngress: 41, ConsEgress: 40},
+		{ConsIngress: 31, ConsEgress: 30},
+		{ConsIngress: 1, ConsEgress: 0},
+	}
+	dpath.Base.PathMeta.CurrHF = 2
+	dpath.HopFields[2].Mac = computeMAC(t, key,
+		dpath.InfoFields[0], dpath.HopFields[2])
+	spkt.Path = dpath
+	payload := []byte("x")
+	buffer := gopacket.NewSerializeBuffer()
+	err := gopacket.SerializeLayers(buffer,
+		gopacket.SerializeOptions{FixLengths: true},
+		spkt, gopacket.Payload(payload))
+	require.NoError(t, err)
+	raw := buffer.Bytes()
+	copy(packetBuffer, raw)
+	tmpBuffer := make([]byte, 4)
+	binary.BigEndian.PutUint32(tmpBuffer, spkt.FlowID)
+	tmpBuffer[0] &= 0xF
+
+	hasher := fnv.New32a()
+	hasher.Write(randomValue)
+	hasher.Write(tmpBuffer[1:4])
+	hasher.Write(packetBuffer[slayers.CmnHdrLen : slayers.CmnHdrLen+spkt.AddrHdrLen()+1])
+	val1 := hasher.Sum32() % uint32(numProcs)
+	val2, err := dp.ComputeProcId(packetBuffer)
```

Comment:
Maybe also test whether two packets of the same flow actually land in the same processing routine?
Also, test whether modifying everything in the packet apart from the packet fields used in the hash input still results in the same hash?